lib/spack/spack/fetch_strategy.py

```diff
@@ -955,87 +952,35 @@ def _clone_src(self) -> None:
-                args.extend(["--branch", self.tag])
-
-            # Try to be efficient if we're using a new enough git.
-            # This checks out only one branch's history
-            if self.git_version >= spack.version.Version("1.7.10"):
-                if self.get_full_repo:
-                    args.append("--no-single-branch")
-                else:
-                    args.append("--single-branch")
-
-            with temp_cwd():
-                # Yet more efficiency: only download a 1-commit deep
-                # tree, if the in-use git and protocol permit it.
-                if (
-                    (not self.get_full_repo)
-                    and self.git_version >= spack.version.Version("1.7.1")
-                    and self.protocol_supports_shallow_clone()
-                ):
-                    args.extend(["--depth", "1"])
-
-                args.extend([self.url])
-                git(*args)
-
-                repo_name = get_single_file(".")
-                if self.stage:
-                    self.stage.srcdir = repo_name
-                shutil.move(repo_name, dest)
-
-            with working_dir(dest):
-                # For tags, be conservative and check them out AFTER
-                # cloning.  Later git versions can do this with clone
-                # --branch, but older ones fail.
-                if self.tag and self.git_version < spack.version.Version("1.8.5.2"):
-                    # pull --tags returns a "special" error code of 1 in
-                    # older versions that we have to ignore.
-                    # see: https://github.com/git/git/commit/19d122b
-                    pull_args = ["pull", "--tags"]
-                    co_args = ["checkout", self.tag]
-                    if not spack.config.get("config:debug"):
-                        pull_args.insert(1, "--quiet")
-                        co_args.insert(1, "--quiet")
+        with temp_cwd():
+            if self.commit:
+                try:
+                    spack.util.git.git_init_fetch(self.url, self.commit, depth, **kwargs)
+                except spack.util.executable.ProcessError:
+                    # TODO(psakeiv) debug message?
+                    spack.util.git.git_clone(self.url, fetch_ref, True, depth, **kwargs)
+            else:
+                spack.util.git.git_clone(self.url, fetch_ref, self.get_full_repo, depth, **kwargs)
+            # if checkout_ref:
+            spack.util.git.git_checkout(checkout_ref, self.git_sparse_paths, **kwargs)
 
-                    git(*pull_args, ignore_errors=1)
-                    git(*co_args)
+            repo_name = get_single_file(".")
+            if self.stage:
+                self.stage.srcdir = repo_name
+            shutil.copytree(repo_name, dest, symlinks=True)
+            shutil.rmtree(
+                repo_name,
+                ignore_errors=False,
+                onerror=fs.readonly_file_handler(ignore_errors=True),
```

Comment:
I feel like this is probably some common operation that could go into `spack.llnl.util.filesystem` if it's not there already, and then we could just wrap this code in a context manager here and not need to have the implementation showing up in the fetch strategy.